doc/man/php-install.1.md
# php-install 1 "Oct 2013" php-install "User Manuals"

## SYNOPSIS

`php-install` [PHP [VERSION]] [-- CONFIGURE_OPTS...]

## DESCRIPTION

Installs different versions of php

https://github.com/marcosdsanchez/php-install#readme

## ARGUMENTS

*PHP*
	Install PHP by name.

*VERSION*
	Optionally select the version of selected PHP.

*CONFIGURE_OPTS*
	Additional optional configure arguments.

## OPTIONS

`-s`, `--src-dir` *DIR*
	Specifies the directory for downloading and unpacking PHP source.

`-i`, `--install-dir` *DIR*
	Specifies the directory where PHP will be installed.
	Defaults to */opt/php_versions/$php-$version* for root and
	*~/.php_versions/$php-$version* for normal users.

`-M`, `--mirror` *URL*
	Specifies an altnerate mirror to download the PHP archive from.

`-u`, `--url` *URL*
	Alternate URL to download the PHP archive from.

`-m`, `--md5` *MD5*
	Specifies the MD5 checksum for the PHP archive.

`--no-download`
	Use the previously downloaded PHP archive.

`--no-verify`
	Do not verify the downloaded PHP archive.

`--no-install-deps`
	Do not install build dependencies before installing PHP.

`--no-reinstall`
	Skip installation if another PHP is detected in same location.

`-V`, `--version`
	Prints the current php-install version.

`-h`, `--help`
	Prints a synopsis of php-install usage.

## EXAMPLES

List supported PHP versions and their major versions:

    $ php-install

Install the current stable version of PHP:

    $ php-install php

Install a latest version of PHP:

    $ php-install php 5.4

Install a specific version of PHP:

    $ php-install php 5.3.14

Install a PHP version into a specific directory:

    $ php-install -i /usr/local/ php 5.3.3

Install a PHP version with specific configuration:

    $ php-install php stable -- --enable-mbstring

## FILES

*/usr/local/src*
	Default root user source directory.
    
*~/src*
	Default non-root user source directory.

*/opt/php_versions/$php-$version*
	Default root user installation directory.

*~/.php_versions/$php-$version*
	Default non-root user installation directory.

## AUTHOR

Marcos D. Sánchez <marcosdsanchez at gmail com>


ChangeLog.md
### 0.0.1 / xxxx-xx-xx
* Initial release:
  * Supports installing arbitrary versions.
  * Supports installing into `/opt/php_versions/` for root and `~/.php_version/` for users
    by default.
  * Supports installing into arbitrary directories.
  * Supports applying arbitrary patches.
  * Supports specifying arbitrary `./configure` options.
  * Supports downloading archives using `wget` or `curl`.
  * Supports verifying downloaded archives using `md5sum` or `md5`.
  * Supports installing build dependencies via the package manager:
    * [apt]
    * [yum]
    * [pacman]
    * [brew]

[apt]: http://wiki.debian.org/Apt
[yum]: http://yum.baseurl.org/
[pacman]: https://wiki.archlinux.org/index.php/Pacman
[brew]: http://mxcl.github.com/homebrew/


share/php-install/php-install.sh
#!/usr/bin/env bash

shopt -s extglob

PHP_INSTALL_VERSION="0.0.1"
PHP_INSTALL_DIR="${BASH_SOURCE[0]%/*}"

PHP_DISTRIBUTIONS=(php)
CONFIGURE_OPTS=()

#
# Auto-detect the package manager.
#
if   command -v apt-get >/dev/null; then PACKAGE_MANAGER="apt"
elif command -v yum     >/dev/null; then PACKAGE_MANAGER="yum"
elif command -v brew    >/dev/null; then PACKAGE_MANAGER="brew"
elif command -v pacman  >/dev/null; then PACKAGE_MANAGER="pacman"
fi

#
# Auto-detect the downloader.
#
if   command -v wget >/dev/null; then DOWNLOADER="wget"
elif command -v curl >/dev/null; then DOWNLOADER="curl"
fi

#
# Auto-detect the md5 utility.
#
if   command -v md5sum >/dev/null; then MD5SUM="md5sum"
elif command -v md5    >/dev/null; then MD5SUM="md5"
fi

#
# Only use sudo if already root.
#
if (( $UID == 0 )); then SUDO=""
else                     SUDO="sudo"
fi

#
# Prints a log message.
#
function log()
{
	if [[ -t 1 ]]; then
		echo -e "\x1b[1m\x1b[32m>>>\x1b[0m \x1b[1m\x1b[37m$1\x1b[0m"
	else
		echo ">>> $1"
	fi
}

#
# Prints a warn message.
#
function warn()
{
	if [[ -t 1 ]]; then
		echo -e "\x1b[1m\x1b[33m***\x1b[0m \x1b[1m\x1b[37m$1\x1b[0m" >&2
	else
		echo "*** $1" >&2
	fi
}

#
# Prints an error message.
#
function error()
{
	if [[ -t 1 ]]; then
		echo -e "\x1b[1m\x1b[31m!!!\x1b[0m \x1b[1m\x1b[37m$1\x1b[0m" >&2
	else
		echo "!!! $1" >&2
	fi
}

#
# Prints an error message and exists with -1.
#
function fail()
{
	error "$*"
	exit -1
}

#
# Searches a file for a key and echos the value.
# If the key cannot be found, the third argument will be echoed.
#
function fetch()
{
	local file="$PHP_INSTALL_DIR/$1.txt"
	local key="$2"
	local pair="$(grep -E "^$key: " "$file")"

	echo "${pair##$key:*( )}"
}

function install_packages()
{
	case "$PACKAGE_MANAGER" in
		apt)	$SUDO apt-get install -y $* ;;
		yum)	$SUDO yum install -y $*     ;;
		brew)
			local brew_owner="$(/usr/bin/stat -f %Su "$(command -v brew)")"
			sudo -u "$brew_owner" brew install $*
			;;
		pacman)
			local missing_pkgs="$(pacman -T $*)"

			if [[ -n "$missing_pkgs" ]]; then
				$SUDO pacman -S $missing_pkgs
			fi
			;;
		"")	warn "Could not determine Package Manager. Proceeding anyways." ;;
	esac
}

#
# Downloads a URL.
#
function download()
{
	local url="$1"
	local dest="$2"

	if [[ -d "$dest" ]]; then
		dest="$dest/${url##*/}"
	fi

	case "$DOWNLOADER" in
		wget) wget -c -O "$dest" "$url"      ;;
		curl) curl -L -C - -o "$dest" "$url" ;;
		"")
			error "Could not find wget or curl"
			return 1
			;;
	esac
}

#
# Verifies a file against a md5 checksum.
#
function verify()
{
	local path="$1"
	local md5="$2"

	if [[ -z "$MD5SUM" ]]; then
		error "Unable to find the md5 checksum utility"
		return 1
	fi

	if [[ -z "$md5" ]]; then
		error "No md5 checksum given"
		return 1
	fi

	if [[ "$($MD5SUM "$path")" != *$md5* ]]; then
		error "$path is invalid!"
		return 1
	fi
}

#
# Extracts an archive.
#
function extract()
{
	local archive="$1"
	local dest="${2:-${archive%/*}}"

	case "$archive" in
		*.tgz|*.tar.gz)		tar -xzf "$archive" -C "$dest" ;;
		*.tbz|*.tbz2|*.tar.bz2)	tar -xjf "$archive" -C "$dest" ;;
		*.zip)			unzip "$archive" -d "$dest" ;;
		*)
			error "Unknown archive format: $archive"
			return 1
			;;
	esac
}

#
# Loads function.sh for the given PHP
#
function load_php()
{
	PHP_DIR="$PHP_INSTALL_DIR/$PHP"

	if [[ ! -d "$PHP_DIR" ]]; then
		echo "php-install: unsupported php: $PHP" >&2
		return 1
	fi

	local expanded_version="$(fetch "$PHP/versions" "$PHP_VERSION")"
	PHP_VERSION="${expanded_version:-$PHP_VERSION}"

	source "$PHP_INSTALL_DIR/functions.sh"
	source "$PHP_DIR/functions.sh"

	PHP_MD5="${PHP_MD5:-$(fetch "$PHP/md5" "$PHP_ARCHIVE")}"
}

#
# Prints PHP versions supported by php-install.
#
function known_php_distributions()
{
	echo "Known php distributions:"

	for php in ${PHP_DISTRIBUTIONS[@]}; do
		echo "  $php:"
		cat "$PHP_INSTALL_DIR/$php/versions.txt" | sed -e 's/^/    /'
	done
}

#
# Prints usage information for php-install.
#
function usage()
{
	cat <<USAGE
usage: php-install [OPTIONS] [PHP [VERSION]] [-- CONFIGURE_OPTS ...]

Options:

	-s, --src-dir DIR	Directory to download source-code into
	-i, --install-dir DIR	Directory to install PHP into
	-m, --md5 MD5		MD5 checksum of the PHP archive
	-u, --url URL       Alternate URL to download the PHP archive from
	--no-download		Use the previously downloaded PHP archive
	--no-verify		Do not verify the downloaded PHP archive
	--no-install-deps	Do not install build dependencies before installing PHP
	--no-reinstall  	Skip installation if another PHP is detected in same location
	-V, --version		Prints the version
	-h, --help		Prints this message

Examples:

	$ php-install php
	$ php-install php 5.3
	$ php-install php 5.4.1
	$ php-install php -- --with-openssl-dir=...

USAGE
}

#
# Parses command-line options for php-install.
#
function parse_options()
{
	local argv=()

	while [[ $# -gt 0 ]]; do
		case $1 in
			-i|--install-dir)
				INSTALL_DIR="$2"
				shift 2
				;;
			-s|--src-dir)
				SRC_DIR="$2"
				shift 2
				;;
			-u|--url)
				PHP_URL="$2"
				shift 2
				;;
			-m|--md5)
				PHP_MD5="$2"
				shift 2
				;;
			--no-download)
				NO_DOWNLOAD=1
				shift
				;;
			--no-verify)
				NO_VERIFY=1
				shift
				;;
			--no-install-deps)
				NO_INSTALL_DEPS=1
				shift
				;;
			--no-reinstall)
				NO_REINSTALL=1
				shift
				;;
			-V|--version)
				echo "php-install: $PHP_INSTALL_VERSION"
				exit
				;;
			-h|--help)
				usage
				exit
				;;
			--)
				shift
				CONFIGURE_OPTS=("$@")
				break
				;;
			-*)
				echo "php-install: unrecognized option $1" >&2
				return 1
				;;
			*)
				argv+=($1)
				shift
				;;
		esac
	done

	case ${#argv[*]} in
		2)
			PHP="${argv[0]}"
			PHP_VERSION="${argv[1]}"
			;;
		1)
			PHP="${argv[0]}"
			PHP_VERSION="stable"
			;;
		0)
			echo "php-install: too few arguments" >&2
			usage 1>&2
			return 1
			;;
		*)
			echo "php-install: too many arguments: ${argv[*]}" >&2
			usage 1>&2
			return 1
			;;
	esac
}
